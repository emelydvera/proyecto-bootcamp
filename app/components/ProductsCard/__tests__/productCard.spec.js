const React = require("react");
const ProductCard = require("../ProductsCard");
const { render, screen } = require("@testing-library/react");
const { getByRole, queryByText, getByLabelText } = screen;
const mockProductWithFreeShipping = require("../../../../mocks/components/productCard/productCardMockWithFreeShipping.json");
const mockProductWithoutFreeShipping = require("../../../../mocks/components/productCardMockWithoutFreeShipping.json");
const mockProductWithoutInstallments = require("../../../../mocks/components/productCardMockWithoutInstallments.json");
const mockProductWithFalsyAddress = require("../../../../mocks/components/productCardMockWithoutAddressState.json");

describe("ProductCard", () => {
  const i18n = { gettext: (text) => text };

  describe("ProuctCard with free shipping", () => {
    let component;
    const props = {
      product: mockProductWithFreeShipping,
      i18n,
      index: 1,
    };

    beforeEach(() => {
      component = render(<ProductCard {...props} />);
    });

    it("should render component", () => {
      const { asFragment } = component;
      expect(asFragment()).toMatchSnapshot();
    });

    it("should render the name of the product", () => {
      const productName = getByRole("heading", { level: 2 });
      expect(productName).toHaveTextContent(mockProductWithFreeShipping.title);
    });

    it("should render the image of the product", () => {
      const productImage = getByRole("img");
      expect(productImage).toHaveAttribute(
        "data-src",
        mockProductWithFreeShipping.thumbnail.replace("http", "https")
      );
    });

    it("should render the city of the product", () => {
      const productCity = queryByText("Antioquia");
      expect(productCity).toBeInTheDocument();
    });

    it("should render the price of the product", () => {
      const productPrice = queryByText(/252900/);
      expect(productPrice).toBeInTheDocument();
    });

    it("should render an anchor tag (<a></a >) with the href containing amount and quantity of the product", () => {
      const productLink = getByRole("link");
      expect(productLink).toHaveAttribute(
        "href",
        `/product/${mockProductWithFreeShipping.id}?quantity=${mockProductWithFreeShipping.installments.quantity}&amount=${mockProductWithFreeShipping.installments.amount}`
      );
    });

    it("should render a tag with 'Envío' when product has free shipping", () => {
      const productFreeShipping = queryByText("Envío");
      expect(productFreeShipping).not.toBeNull();
    });
  });

  describe("ProuctCard without free shipping", () => {
    let component;
    const props = {
      product: mockProductWithoutFreeShipping,
      i18n,
      index: 1,
    };

    beforeEach(() => {
      component = render(<ProductCard {...props} />);
    });

    it("should render component", () => {
      const { asFragment } = component;
      expect(asFragment()).toMatchSnapshot();
    });

    it("should render the name of the product", () => {
      const productName = getByRole("heading", { level: 2 });
      expect(productName).toHaveTextContent(
        mockProductWithoutFreeShipping.title
      );
    });

    it("should render the city of the product", () => {
      const productCity = queryByText("Bolívar");
      expect(productCity).toBeInTheDocument();
    });

    it("should render the price of the product", () => {
      const productPrice = queryByText(/474999/);
      expect(productPrice).toBeInTheDocument();
    });

    it("should render the image of the product", () => {
      const productImage = getByRole("img");
      expect(productImage).toHaveAttribute(
        "data-src",
        mockProductWithoutFreeShipping.thumbnail.replace("http", "https")
      );
    });

    it("should render an anchor tag (<a></a >) with the href containing amount and quantity of the product", () => {
      const productLink = getByRole("link");
      expect(productLink).toHaveAttribute(
        "href",
        `/product/${mockProductWithoutFreeShipping.id}?quantity=${mockProductWithFreeShipping.installments.quantity}&amount=${mockProductWithFreeShipping.installments.amount}`
      );
    });

    it("should not render a tag with 'Envío' when product has no free shipping", () => {
      const productFreeShipping = queryByText("Envío");
      expect(productFreeShipping).toBeNull();
    });
  });

  describe("ProuctCard without installments", () => {
    let component;
    const props = {
      product: mockProductWithoutInstallments,
      i18n,
      index: 1,
    };

    beforeEach(() => {
      component = render(<ProductCard {...props} />);
    });

    it("should render component", () => {
      const { asFragment } = component;
      expect(asFragment()).toMatchSnapshot();
    });

    it("should render the name of the product", () => {
      const productName = getByRole("heading", { level: 2 });
      expect(productName).toHaveTextContent(
        mockProductWithoutInstallments.title
      );
    });

    it("should render the city of the product", () => {
      const productCity = queryByText("Cundinamarca");
      expect(productCity).toBeInTheDocument();
    });

    it("should render the price of the product", () => {
      const productPrice = queryByText(/400000/);
      expect(productPrice).toBeInTheDocument();
    });

    it("should render the image of the product", () => {
      const productImage = getByRole("img");
      expect(productImage).toHaveAttribute(
        "data-src",
        mockProductWithoutInstallments.thumbnail.replace("http", "https")
      );
    });

    it("should render an anchor tag (<a></a >) with the href not containing amount and quantity of the product", () => {
      const productLink = getByRole("link");
      expect(productLink).toHaveAttribute(
        "href",
        `/product/${mockProductWithoutInstallments.id}`
      );
    });

    it("should not render a tag with 'Envío' when product has no free shipping", () => {
      const productFreeShipping = queryByText("Envío");
      expect(productFreeShipping).toBeNull();
    });
  });

  describe("ProuctCard with state_name property empty or undefined", () => {
    let component;
    const props = {
      product: mockProductWithFalsyAddress,
      i18n,
      index: 1,
    };

    beforeEach(() => {
      component = render(<ProductCard {...props} />);
    });

    it("should render component", () => {
      const { asFragment } = component;
      expect(asFragment()).toMatchSnapshot();
    });

    it("should render the name of the product", () => {
      const productName = getByRole("heading", { level: 2 });
      expect(productName).toHaveTextContent(mockProductWithFalsyAddress.title);
    });

    it("should render 'Sin ubicación' when the city of is not available in the product", () => {
      const productCity = getByLabelText("Sin ubicación");
      expect(productCity).toBeInTheDocument();
    });

    it("should render the price of the product", () => {
      const productPrice = queryByText(/400000/);
      expect(productPrice).toBeInTheDocument();
    });

    it("should render the image of the product", () => {
      const productImage = getByRole("img");
      expect(productImage).toHaveAttribute(
        "data-src",
        mockProductWithFalsyAddress.thumbnail.replace("http", "https")
      );
    });
  });
});
