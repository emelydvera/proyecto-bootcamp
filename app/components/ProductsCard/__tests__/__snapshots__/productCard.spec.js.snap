// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductCard ProuctCard with free shipping should render component 1`] = `
<DocumentFragment>
  <article
    class="andes-card andes-card--flat andes-card--default andes-card--padding-default"
  >
    <div
      class="andes-card__content card"
    >
      <div
        class="card__image"
      >
        <figure>
          <a
            href="/product/MCO640590627?quantity=12&amount=315833.33"
          >
            <img
              alt="imagen del producto"
              class="card__image__img lazy-loadable"
              data-src="https://http2.mlstatic.com/D_666275-MCO48851447655_012022-I.jpg"
              decoding="async"
              src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
              tabindex="25"
            />
          </a>
        </figure>
      </div>
      <div
        class="card__description"
      >
        <div
          class="card__description__main-info"
        >
          <div
            aria-label="precio del producto: \${0}, {1}"
            class="money__and__shipping"
            tabindex="24"
          >
            <span
              class="andes-money-amount price andes-money-amount--cents-superscript"
              style="font-size: 24px;"
            >
              <span
                class="andes-visually-hidden"
              >
                252900 pesos
              </span>
              <span
                aria-hidden="true"
                class="andes-money-amount__currency-symbol"
              >
                $
              </span>
              <span
                aria-hidden="true"
                class="andes-money-amount__fraction"
              >
                252.900
              </span>
              <span
                aria-hidden="true"
                class="andes-visually-hidden"
              >
                ,
              </span>
              <span
                aria-hidden="true"
                class="andes-money-amount__cents andes-money-amount__cents--superscript-24"
                style="font-size: 12px; margin-top: 4px;"
              >
                00
              </span>
            </span>
            <span
              class="shipping"
            >
              Envío
            </span>
          </div>
          <h2
            class="title"
            tabindex="23"
          >
            Cargador De Repuesto Para Mac Book, Cargador De Computadora
          </h2>
        </div>
        <span
          aria-label="Ubicación del producto: {0}"
          class="card__description__location"
          tabindex="24"
        >
          Antioquia
        </span>
      </div>
    </div>
  </article>
</DocumentFragment>
`;

exports[`ProductCard ProuctCard without free shipping should render component 1`] = `
<DocumentFragment>
  <article
    class="andes-card andes-card--flat andes-card--default andes-card--padding-default"
  >
    <div
      class="andes-card__content card"
    >
      <div
        class="card__image"
      >
        <figure>
          <a
            href="/product/MCO640590627?quantity=12&amount=315833.33"
          >
            <img
              alt="imagen del producto"
              class="card__image__img lazy-loadable"
              data-src="https://http2.mlstatic.com/D_666275-MCO48851447655_012022-I.jpg"
              decoding="async"
              src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
              tabindex="25"
            />
          </a>
        </figure>
      </div>
      <div
        class="card__description"
      >
        <div
          class="card__description__main-info"
        >
          <div
            aria-label="precio del producto: \${0}, {1}"
            class="money__and__shipping"
            tabindex="24"
          >
            <span
              class="andes-money-amount price andes-money-amount--cents-superscript"
              style="font-size: 24px;"
            >
              <span
                class="andes-visually-hidden"
              >
                4749999 pesos
              </span>
              <span
                aria-hidden="true"
                class="andes-money-amount__currency-symbol"
              >
                $
              </span>
              <span
                aria-hidden="true"
                class="andes-money-amount__fraction"
              >
                4.749.999
              </span>
              <span
                aria-hidden="true"
                class="andes-visually-hidden"
              >
                ,
              </span>
              <span
                aria-hidden="true"
                class="andes-money-amount__cents andes-money-amount__cents--superscript-24"
                style="font-size: 12px; margin-top: 4px;"
              >
                00
              </span>
            </span>
          </div>
          <h2
            class="title"
            tabindex="23"
          >
            Apple Macbook Air (13 Pulgadas, 2020, Chip M1, 256 Gb De Ssd, 8 Gb De Ram) - Gris Espacial
          </h2>
        </div>
        <span
          aria-label="Ubicación del producto: {0}"
          class="card__description__location"
          tabindex="24"
        >
          Bolívar
        </span>
      </div>
    </div>
  </article>
</DocumentFragment>
`;

exports[`ProductCard ProuctCard without installments should render component 1`] = `
<DocumentFragment>
  <article
    class="andes-card andes-card--flat andes-card--default andes-card--padding-default"
  >
    <div
      class="andes-card__content card"
    >
      <div
        class="card__image"
      >
        <figure>
          <a
            href="/product/MCO640590621"
          >
            <img
              alt="imagen del producto"
              class="card__image__img lazy-loadable"
              data-src="https://http2.mlstatic.com/D_666275-MCO48851447655_012022-I.jpg"
              decoding="async"
              src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
              tabindex="25"
            />
          </a>
        </figure>
      </div>
      <div
        class="card__description"
      >
        <div
          class="card__description__main-info"
        >
          <div
            aria-label="precio del producto: \${0}, {1}"
            class="money__and__shipping"
            tabindex="24"
          >
            <span
              class="andes-money-amount price andes-money-amount--cents-superscript"
              style="font-size: 24px;"
            >
              <span
                class="andes-visually-hidden"
              >
                400000 pesos
              </span>
              <span
                aria-hidden="true"
                class="andes-money-amount__currency-symbol"
              >
                $
              </span>
              <span
                aria-hidden="true"
                class="andes-money-amount__fraction"
              >
                400.000
              </span>
              <span
                aria-hidden="true"
                class="andes-visually-hidden"
              >
                ,
              </span>
              <span
                aria-hidden="true"
                class="andes-money-amount__cents andes-money-amount__cents--superscript-24"
                style="font-size: 12px; margin-top: 4px;"
              >
                00
              </span>
            </span>
          </div>
          <h2
            class="title"
            tabindex="23"
          >
            Chevrolet Captiva Sport At 3.0 2012
          </h2>
        </div>
        <span
          aria-label="Ubicación del producto: {0}"
          class="card__description__location"
          tabindex="24"
        >
          Cundinamarca
        </span>
      </div>
    </div>
  </article>
</DocumentFragment>
`;
